### PR TITLE
Redis sentinel support

### PR DESCRIPTION
**Pull request by Apollon77**
This PR is about to add redis-sentinel support for objects and state DBs including the needed testing for it.
The following is in:
* allow to specify one host for states/objects or a comma-separated-list of hosts or hosts as array of strings
* allow same for ports, if multiple hosts are set but only one port number then the same port is used for all host entries
* when we have a list of hosts initialize the ioredis connection with those as sentinels. Then ioredis will find the current master and connects to them and also handle reconnects on master failovers
* setup custom commnd also allows to enter comma separated hosts and ports

### CODE REVIEW

**Comment by Apollon77** [comment]
All finished. testing works (https://travis-ci.org/ioBroker/ioBroker.js-controller/jobs/561218820#L5037) incl failover test (https://travis-ci.org/ioBroker/ioBroker.js-controller/jobs/561218820#L5988). And objects ha lib updated

**Comment by Apollon77** [comment]
@AlCalzone @GermanBluefox Review would be much appereciated!!!

**Comment by GermanBluefox** [comment]
lgtm

File: .travis.yml [hunk 1]
```diff
@@ -2,15 +2,17 @@ os:
...
-services:
-  - redis-server
 language: node_js
 node_js:
   - '6'
   - '8'
   - '10'
   - '12'
 before_install:
+  - 'if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then sudo add-apt-repository ppa:chris-lea/redis-server -y; fi'
```
Comment (AlCalzone) [inline .travis.yml]: Geht glaube ich auch etwas schöner: https://docs.travis-ci.com/user/installing-dependencies/#installing-packages-with-the-apt-addon

File: conf/sources-dist.json [hunk 2]
```diff
@@ -10,10 +10,9 @@
   "alexa2": {
```
Comment (AlCalzone) [inline conf/sources-dist.json]: If you ask me, this file change should be part of a separate PR

File: lib/setup/setupSetup.js [hunk 3]
```diff
@@ -217,42 +220,56 @@ function Setup(options) {
                     op = 9001;
                 } else if (otype === 'redis') {
                     op = 6379;
+                    if (ohost.includes(',')) {
+                        ohost = ohost.split(',');
+                        ohost.forEach((host, idx) => {
+                            ohost[idx] = host.trim();
+                        });
+                        op = 26379;
```
Comment (AlCalzone) [inline lib/setup/setupSetup.js]: I like speaking names, especially since below it was called `oport`.

File: lib/setup/setupSetup.js [hunk 4]
```diff
@@ -217,42 +220,56 @@ function Setup(options) {
...
-                        if (isNaN(oport)) {
-                            console.log('Invalid objects port: ' + oport);
-                            callback(23);
+                        if (userPort.includes(',')) {
+                            oport = userPort.split(',');
+                            oport.forEach((port, idx) => {
+                                oport[idx] = parseInt(port.trim(), 10);
+                                if (isNaN(oport[idx])) {
+                                    console.log('Invalid objects port: ' + oport[idx]);
+                                    callback(23);
```
Comment (AlCalzone) [inline lib/setup/setupSetup.js]: I think this construct (forEach plus callback) can result in the callback method being called multiple times. For process.exit that doesn't matter, but it is possible that for some application callback is NOT `process.exit`.

I believe this should be a classic for loop with a `break` statement (or even `return`) after `callback(23)`.

File: lib/setup/setupSetup.js [hunk 5]
```diff
@@ -217,42 +220,56 @@ function Setup(options) {
...
+                        } else {
+                            oport = parseInt(userPort, 10);
+                            if (isNaN(oport)) {
+                                console.log('Invalid objects port: ' + oport);
+                                callback(23);
+                            }
                         }
                     }
-                    rl.question('Type of states DB [(f)file, (r)edis], default [' + ot + ']: ', stype => {
+                    rl.question('Type of states DB [(f)file, (r)edis], default [' + otype + ']: ', stype => {
```
Comment (AlCalzone) [inline lib/setup/setupSetup.js]: Thank you! :)

File: lib/setup/setupSetup.js [hunk 6]
```diff
@@ -264,21 +281,35 @@ function Setup(options) {
                                 sp = 9000;
                             } else if (stype === 'redis') {
                                 sp = 6379;
+                                if (shost.includes(',')) {
+                                    shost = shost.split(',');
+                                    shost.forEach((host, idx) => {
+                                        shost[idx] = host.trim();
+                                    });
```
Comment (AlCalzone) [inline lib/setup/setupSetup.js]: `shost = shost.split(',').map(host => host.trim());`

File: lib/setup/setupSetup.js [hunk 7]
```diff
@@ -264,21 +281,35 @@ function Setup(options) {
...
-                                    if (isNaN(sport)) {
-                                        console.log('Invalid states port: ' + sport);
-                                        callback(23);
+                                    if (userPort.includes(',')) {
+                                        sport = userPort.split(',');
+                                        sport.forEach((port, idx) => {
+                                            sport[idx] = parseInt(port.trim(), 10);
+                                            if (isNaN(sport[idx])) {
+                                                console.log('Invalid states port: ' + sport[idx]);
+                                                callback(23);
```
Comment (AlCalzone) [inline lib/setup/setupSetup.js]: See above

File: lib/states/statesInRedis.js [hunk 8]
```diff
@@ -105,12 +105,27 @@ class StateRedis {
...
+            this.log.debug('Redis States: Use File Socket for connection: ' + this.settings.connection.options.path);
+        } else if (Array.isArray(this.settings.connection.host)) { // Host is an array means we use a sentinel
+            const defaultPort = Array.isArray(this.settings.connection.port) ? null : this.settings.connection.port;
+            this.settings.connection.options.sentinels = [];
+            this.settings.connection.host.forEach((redisNode, idx) => {
+                this.settings.connection.options.sentinels.push({
+                    host: redisNode,
+                    port: defaultPort ? defaultPort : this.settings.connection.port[idx]
+                });
+            });
```
Comment (AlCalzone) [inline lib/states/statesInRedis.js]: `this.settings.connections.options.sentinels = this.settings.connections.host.map((redisNode, idx) => ({ ... }));`

File: lib/states/statesInRedis.js [hunk 9]
```diff
@@ -105,12 +105,27 @@ class StateRedis {
...
+            // initiate a unix socket connection
+            this.settings.connection.options.path = this.settings.connection.host;
+            this.log.debug('Redis States: Use File Socket for connection: ' + this.settings.connection.options.path);
+        } else if (Array.isArray(this.settings.connection.host)) { // Host is an array means we use a sentinel
+            const defaultPort = Array.isArray(this.settings.connection.port) ? null : this.settings.connection.port;
+            this.settings.connection.options.sentinels = [];
+            this.settings.connection.host.forEach((redisNode, idx) => {
+                this.settings.connection.options.sentinels.push({
+                    host: redisNode,
+                    port: defaultPort ? defaultPort : this.settings.connection.port[idx]
```
Comment (AlCalzone) [inline lib/states/statesInRedis.js]: what about `port: defaultPort || ...`

File: test/redis-sentinel/log/.gitignore [hunk 10]
```diff
@@ -0,0 +1 @@
+./*.log
```
Comment (AlCalzone) [inline test/redis-sentinel/log/.gitignore]: Could all these .gitignores be merged into a single one in the root dir?

File: test/redis-sentinel/testStatesRedis.js [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+/* jshint strict:false */
+/* jslint node:true */
+/* jshint expr:true */
+'use strict';
+
+const expect  = require('chai').expect;
+const setup   = require(__dirname + '/../lib/setup4controller');
+const fs      = require('fs');
+const { exec } = require('child_process');
+const dataDir = __dirname + '/../../tmp/data';
```
Comment (AlCalzone) [inline test/redis-sentinel/testStatesRedis.js]: `path.join` pls

File: test/redis-sentinel/testStatesRedis.js [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+let objects     = null;
+let states      = null;
+let onStatesChanged = null;
+
+function cleanDbs() {
+    if (fs.existsSync(dataDir + '/objects.json')) {
```
Comment (AlCalzone) [inline test/redis-sentinel/testStatesRedis.js]: dito

File: .travis.yml (hunk 1 above)
Comment (Apollon77) [reply .travis.yml]: I exactly got that soluton from the same page more up https://docs.travis-ci.com/user/installing-dependencies/#installing-packages-from-a-custom-apt-repository

It was not working else when I remember

File: conf/sources-dist.json (hunk 2 above)
Comment (Apollon77) [reply conf/sources-dist.json]: Yes don't know why it went into it

File: lib/setup/setupSetup.js (hunk 3 above)
Comment (Apollon77) [reply lib/setup/setupSetup.js]: I just used same namings as it was before and did minimal changes only ;-)

File: lib/setup/setupSetup.js (hunk 5 above)
Comment (Apollon77) [reply lib/setup/setupSetup.js]: yw

**Comment by Apollon77** [comment]
Thank you, will check tomorrow

File: lib/setup/setupSetup.js (hunk 4 above)
Comment (Apollon77) [reply lib/setup/setupSetup.js]: I added it, but in fact in standard case "callback " will be process exit :-)

File: lib/setup/setupSetup.js (hunk 6 above)
Comment (Apollon77) [reply lib/setup/setupSetup.js]: used, I still need tzo think into these style of code :-)

File: lib/states/statesInRedis.js (hunk 8 above)
Comment (Apollon77) [reply lib/states/statesInRedis.js]: changed, hopefully correct

File: lib/states/statesInRedis.js (hunk 9 above)
Comment (Apollon77) [reply lib/states/statesInRedis.js]: fixed

File: test/redis-sentinel/log/.gitignore (hunk 10 above)
Comment (Apollon77) [reply test/redis-sentinel/log/.gitignore]: hm ... it is the question ... splitting is more special and so you know why things are where ...

File: test/redis-sentinel/testStatesRedis.js (hunk 11 above)
Comment (Apollon77) [reply test/redis-sentinel/testStatesRedis.js]: done (file was just copied)

File: conf/sources-dist.json (hunk 2 above)
Comment (Apollon77) [reply conf/sources-dist.json]: reverted

File: lib/states/statesInRedis.js [hunk 13]
```diff
@@ -112,12 +112,10 @@ class StateRedis {
         } else if (Array.isArray(this.settings.connection.host)) { // Host is an array means we use a sentinel
             const defaultPort = Array.isArray(this.settings.connection.port) ? null : this.settings.connection.port;
             this.settings.connection.options.sentinels = [];
```
Comment (AlCalzone) [inline lib/states/statesInRedis.js]: That line is now unnecessary

File: test/redis-sentinel/testStatesRedis.js [hunk 14]
```diff
@@ -4,27 +4,28 @@
 /* jshint expr:true */
 'use strict';
 
-const expect  = require('chai').expect;
-const setup   = require(__dirname + '/../lib/setup4controller');
 const fs      = require('fs');
+const path    = require('path');
+const expect  = require('chai').expect;
+const setup   = require(path.join(__dirname,'..', 'lib', 'setup4controller'));
```
Comment (AlCalzone) [inline test/redis-sentinel/testStatesRedis.js]: you didn't have to split it all up, but fine with me :)

File: lib/states/statesInRedis.js (hunk 13 above)
Comment (Apollon77) [reply lib/states/statesInRedis.js]: changed

File: test/redis-sentinel/testStatesRedis.js (hunk 14 above)
Comment (Apollon77) [reply test/redis-sentinel/testStatesRedis.js]: if then correct ;-)

File: lib/setup/setupSetup.js (hunk 4 above)
Comment (AlCalzone) [reply lib/setup/setupSetup.js]: Ok lets deal with it when we get a bug report ;)

Comment (Apollon77) [reply lib/setup/setupSetup.js]: :-)
